src/test/java/gov/usgs/earthquake/aws/JsonNotificationIndexTest.java

```diff
@@ -0,0 +1,531 @@
+    t1.start();
+    t2.start();
+
+    // Wait for the inserts to finish
+    t1.join();
+    t2.join();
+
+    Assert.assertEquals(SOME_NOTIFICATIONS.size()
+        + EXPIRED_NOTIFICATIONS.size() + 4, getIndexSize());
+  }
+
+  // ------------------------------------------------------------------------
+  // Helper/Utility Methods
+  // ------------------------------------------------------------------------
+  private boolean exists(Notification n) throws Exception {
+    List<Notification> idx = index
+        .getNotifications(_query_allNotifications);
+
+    return contains(idx, n);
+  }
+
+  private boolean contains(List<Notification> haystack, Notification needle) {
+    for (int i = 0; i < haystack.size(); ++i) {
+      if (haystack.get(i).equals(needle)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private int getIndexSize() throws Exception {
+    int i = 0;
+    ResultSet rs = connection.createStatement().executeQuery(
+        "SELECT * FROM "
+        + JsonNotificationIndex.DEFAULT_TABLE);
+
+    while (rs.next()) {
+      ++i;
+    }
+    rs.close();
+
+    return i;
+  }
+
+  private static Notification createExpiredNotification() {
+
+    Notification n;
+    Date d = new Date((new Date()).getTime() - 1);
+    ProductId id = new ProductId("test", "expired", String.valueOf(Math
+        .rint(Math.random() * 1000000)));
+
+    // Just a "randomization" to sometimes create URL vs. Default types
+    if ((d.getTime() % 2) == 0) {
+      // Even; so create a URL type
+      try {
+        n = new URLNotification(id, d, null, new URL(
+            "http://earthquake.usgs.gov/expired/" + id.getCode()));
+      } catch (MalformedURLException mfe) {
+        // Oh well, it doesn't really matter, just use a default
+        // notification instead.
+        n = new DefaultNotification(id, d, null);
+      }
+    } else {
+      // Odd; so create a default type
+      n = new DefaultNotification(id, d, null);
+    }
+
+    return n;
+  }
+
+  private static Notification createRandomNotification(String source,
+      String type) {
+
+    Notification n = null;
+    ProductId id = new ProductId(source, type, String.valueOf(Math
+        .rint(Math.random() * 1000000)));
+    String url = "http://earthquake.usgs.gov/pdtester/" + id.getCode();
+    Date d = null;
+
+    try {
+      long time = (long) (new Date()).getTime()
+          + (long) Math.rint(Math.random() * 1000000000);
+      d = new Date(time);
+
+      if ((time % 2) == 0) {
+        n = new URLNotification(id, d, null, new URL(url));
+      } else {
+        n = new DefaultNotification(id, d, null);
+      }
+
+    } catch (MalformedURLException mfe) {
+      n = new DefaultNotification(id, d, null);
+    }
+
+    return n;
+  }
+
+  private class SimpleInsertThread extends Thread {
+    public void run() {
+      try {
+        index.addNotification(TEST_DEFAULT_NOTIFICATION);
+        Assert.assertTrue(exists(TEST_DEFAULT_NOTIFICATION));
```

Comment:
Asserting inside the sub-thread won't fail the test. Should probably refactor.

Reply:
refactored here, and in original it was based on